Serialization/json.h

```diff
@@ -128,10 +131,130 @@ std::string JSON(const T& value) {
+  virtual ~ParseJSONException() {}
+};
+
+struct JSONSchemaException : ParseJSONException {
+  // TODO(dkorolev): Eventually, only trace and dump `value` with full path in debug builds.
+  const std::string expected_;
+  const std::string value_;
+  JSONSchemaException(const std::string& expected, rapidjson::Value& value, const std::string& path)
+      : expected_(expected), value_(NonThrowingFormatRapidJSONValueAsString(value, path.substr(1))) {}
+  static std::string NonThrowingFormatRapidJSONValueAsString(rapidjson::Value& value, const std::string& path) {
+    // Attempt to generate a human-readable description of the part of the JSON,
+    // that has been parsed but is of wrong schema.
+    try {
+      std::ostringstream os;
+      auto stream = rapidjson::GenericWriteStream(os);
+      auto writer = rapidjson::Writer<rapidjson::GenericWriteStream>(stream);
+      rapidjson::Document document;
+      document.SetObject();
+      document.AddMember(path.c_str(), value, document.GetAllocator());
+      document.Accept(writer);
+      return os.str();
+    } catch (const std::exception& e) {
+      return "The `" + path + "` field could not be parsed.";
+    }
+  }
+  virtual const char* what() const throw() { return ("Expected " + expected_ + ", got: " + value_).c_str(); }
+};
+
+struct InvalidJSONException : ParseJSONException {
+  const std::string erroneus_json_;
+  explicit InvalidJSONException(const std::string& json) : erroneus_json_(json) {}
+  virtual const char* what() const throw() { return erroneus_json_.c_str(); }
+};
+
+// TODO(dkorolev): A smart `enable_if` to not treat any non-primitive type as a `CURRENT_STRUCT`?
+template <typename T>
+struct LoadFromJSONImpl {
+  struct LoadFieldVisitor {
+    rapidjson::Value& source_;
+    const std::string& path_;
+
+    explicit LoadFieldVisitor(rapidjson::Value& source, const std::string& path)
+        : source_(source), path_(path) {}
+
+    // IMPORTANT: Pass in `const char* name`, as `const std::string& name`
+    // would fail memory-allocation-wise due to over-smartness of RapidJSON.
+    template <typename U>
+    void operator()(const char* name, U& value) const {
+      LoadFromJSONImpl<U>::Load(source_[name], value, path_ + '.' + name);
+    }
+  };
+
+  static void Load(rapidjson::Value& source, T& destination, const std::string& path) {
+    if (!source.IsObject()) {
+      throw JSONSchemaException("object", source, path);
+    }
+    LoadFieldVisitor visitor(source, path);
+    current::reflection::VisitAllFields<bricks::decay<T>,
+                                        current::reflection::FieldNameAndMutableValue>::WithObject(destination,
+                                                                                                   visitor);
+  }
+};
+template <>
+struct LoadFromJSONImpl<uint64_t> {
+  static void Load(rapidjson::Value& source, uint64_t& destination, const std::string& path) {
+    if (!source.IsNumber()) {
```

Comment:
Very good point. I digged further, into `rapidjson/document.h`:

```
// By using proper binary layout, retrieval of different integer types do not need conversions.
union Number {
...
```

I read it that `IsNumber()` is the right check, and `GetUint64()` is the right getter for `uint64_t`.
Re, `float` and `double`, there is no `float` in RapidJSON.

Thus, it appears the code in place now does the right thing.